### PR TITLE
Introduce WST admin sub-team

### PR DESCRIPTION
**Pull request by suushiemaniac**
Adds a new `wst_admin` sub-team for more fine-grained controls of admin privileges. This change is motivated by the onboarding of new members around May 2021, where we don't want to give them instant access to the entire infrastructure.

All admin privileges that have previously been determined by WST membership are now transferred to this new sub-team. On staging, admin privileges continue to be deterimined by "general" WST membership to facilitate development and testing for new members.

As a bonus feature, we could also use this admin team independently from the rest of the WST, for example to include other team leaders who need certain privileges. This is currently **not** implemented but just a perk that is possible.

Open points for feedback:
- Is the distinction between prod and staging for different admin levels (see above) reasonable?
- Is the ternary check in `user.rb` correct? Having two question marks feels awkward.
- Who creates the `software-admin@worldcubeassociation.org` mail address? I refrained from linking the team to `admin@` on purpose, in order to give us more degrees of freedom to decide which inbox forwards to where without breaking existing structures.

# IMPORTANT NOTE
This feature is not yet "armed", in order to avoid locking ourselves out. We need to manually populate the teams on the website and on GitHub and then replace the current checks with the commented-out lines directly below to enable syncs.

### CODE REVIEW

**Comment by Jambrose777** [comment]
> * Is the ternary check in `user.rb` correct? Having two question marks feels awkward.

Yes that's correct, its used kinda often in our codebase already.

> This feature is not yet "armed", in order to avoid locking ourselves out. We need to manually populate the teams on the website and on GitHub and then replace the current checks with the commented-out lines directly below to enable syncs.

One thing that we can do is have The Board or WRT populate the sub-team right after its live. But I also see how that's still not ideal to have all admins locked out even for a moment.

File: chef/site-cookbooks/wca/templates/gen-authorized-keys.sh.erb [hunk 1]
```diff
@@ -14,6 +14,8 @@ fi
   --gh-org thewca \
   --gh-team "Software Team" \
   --keyfile /home/cubing/.ssh/authorized_keys
+# TODO GB enable once the software admin team has been populated on the website
```
Comment (jonatanklosko) [inline chef/site-cookbooks/wca/templates/gen-authorized-keys.sh.erb:17]: To clarify: this loads the SSH keys from GitHub, so we just need to populate the GitHub team, not the website one.

**Review by jonatanklosko — COMMENTED** [review]
Thanks for taking this on! I should've really done it, so sorry for delaying on that :<

You also need to update the `config/locales/en.yml` file with the team name (see #6058). It's hidden, but the name is used on the admin teams page.

---

I believe the course of actions would be as follows (please correct me if I miss something):

* populate the GH admin team and test the `gen-authorized-keys.sh.erb` script on the staging server (i.e. after running chef, there should be proper keys in `.ssh/authorized_keys`)
* create the GSuite admin group/email
* merge this PR, keeping `User#admin?` unchanged
* populate admins on the website
* send a follow up PR with the updated `User#admin?`

One thing to keep in mind is that the `software@` members have access to the WST drive, so we should make a separate drive for admins and move relevant files there (definitely the current secrets file, but perhaps there are other confidential ones).

**Comment by suushiemaniac** [comment]
> Yes that's correct, its used kinda often in our codebase already.

Thanks! Then I guess it's just my intuition not being accustomed to Ruby syntax 😅 

> One thing that we can do is have The Board or WRT populate the sub-team right after its live. But I also see how that's still not ideal to have all admins locked out even for a moment.

Yes, that's the point. Also, since we are making things even more restrictive than they currently are, I figured this would be the best way to "dry-run" this feature before actually adding new members.

File: chef/site-cookbooks/wca/templates/gen-authorized-keys.sh.erb (hunk 1 above)
Comment (suushiemaniac) [reply chef/site-cookbooks/wca/templates/gen-authorized-keys.sh.erb:17]: Yes, that was just me not paying attention while copy/pasting my TODO comment from `user.rb`. Nice catch!

**Comment by suushiemaniac** [comment]
> Thanks for taking this on! I should've really done it, so sorry for delaying on that :<

Absolutely no need to worry! The (long-term) purpose of this commit is to allow onboarding of more members, which should ideally also help to reduce our singular dependency on you for coding website changes. Would be kind of ironic if we'd have impose on you to code a feature that exists so that we don't have to impose on you any more.

> You also need to update the `config/locales/en.yml` file with the team name (see #6058). It's hidden, but the name is used on the admin teams page.

Alright, will do!

> I believe the course of actions would be as follows (please correct me if I miss something):
> 
> * populate the GH admin team and test the `gen-authorized-keys.sh.erb` script on the staging server (i.e. after running chef, there should be proper keys in `.ssh/authorized_keys`)
> * create the GSuite admin group/email
> * merge this PR, keeping `User#admin?` unchanged
> * run chef on production to update the SSH keys
> * populate admins on the website
> * send a follow up PR with the updated `User#admin?`

Sounds good to me, but others feel free to chime in and double-check (or triple-check) to make sure.

> One thing to keep in mind is that the `software@` members have access to the WST drive, so we should make a separate drive for admins and move relevant files there (definitely the current secrets file, but perhaps there are other confidential ones).

Yes, this can be done in conjunction with the creation of the new `software-admin@` account.

**Comment by suushiemaniac** [comment]
I also noticed quite a lot of tests are failing, some of which seem to be related to the changes introduced in this PR. Unfortunately I won't have time to investigate before Friday. So if anyone else from WST wants to contribute, I've enabled the "Allow edits by maintainers" feature which means you can push to my fork at the `feature/wst-admin` branch of this PR.

**Comment by suushiemaniac** [comment]
Friday came early! Running the rspec tests locally results in a cascade calamity of failures, because the pages that Capybara receives are rendered in German but the tests all assume English.

How can I configure the local testing JS engine to render pages in English? @jonatanklosko

**Comment by suushiemaniac** [comment]
Closing and re-opening in an attempt to convince Travis to check this PR

**Comment by jonatanklosko** [comment]
> How can I configure the local testing JS engine to render pages in English? @jonatanklosko

Hmm, never heard anyone had this case. IIRC the server tries to guess the appropriate locale from the system-wide one, so perhaps that is the case in tests as well. You can check if modifying the global [`set_locale`](https://github.com/thewca/worldcubeassociation.org/blob/997e519371cf3758c3d72fca768edfbbd37733f3/WcaOnRails/app/controllers/application_controller.rb#L21-L30) middleware to ignore `http_accept_language` would help.

> Closing and re-opening in an attempt to convince Travis to check this PR

I think this is a more broad problem with Travis. We migrated from travis.org to travis.com as they will be shutting down the former soon. Supposedly this shouldn't have impacted open source projects, but it turns out that after running out of the default credits (which seems we did), we need to contact Travis as they work on case-by-case basis... ([ref](https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works)). Ideally we would migrate to GH actions.

**Comment by suushiemaniac** [comment]
Thanks for the feedback! I have "solved" the problem by hard-coding `I18n.locale = :en` in the `set_locale` method and indeed all tests succeed, except for the two or three that deliberately change the locale to Spanish.

Now that all tests check out on my machine, how should we proceed?

**Comment by jonatanklosko** [comment]
A more robust option you could try is modifying the `set_locale` method to  do something like this:

```rb
session[:locale] ||= current_user&.preferred_locale

unless Rails.env.test? do
  session[:locale] ||= http_accept_language.language_region_compatible_from(I18n.available_locales)
end
```

This way in tests we would ignore any system information. Feel free to try this out, and if it works include in the PR.

---

> Now that all tests check out on my machine, how should we proceed?

Migrating the build is not as straightforward to block this PR, so we should remove the required build check and proceed as per the list above. Before merging the PR we should 1. test the authorized keys on the staging and 2. create the GSuite group

**Comment by suushiemaniac** [comment]
> This way in tests we would ignore any system information. Feel free to try this out, and if it works include in the PR.
 
I can already see that this won't work because `sign_up_spec.rb` tests the localization by deliberately setting the `Accept-Language` header and expecting the website to react accordingly.

In my opinion, the preferred way would be to tell Capybara/Poltergeist/whatever to send the `en-EN` headers by default on every request (unless explicitly overridden in the individual test case). It seems [this is possible](https://github.com/teamcapybara/capybara#racktest) with the default RackTest driver, but I can't figure out how to handle it with our Poltergeist driver (which also seems deprecated judging by the fact that its repository is deprecated?). We should proceed without wasting more time figuring out this specific pain point on my local machine.

**Comment by suushiemaniac** [comment]
Manual tests on staging, including the SSH key sync, have been successful. I am now deploying this feature to prod in its current "unarmed" state

**Pull request merged**